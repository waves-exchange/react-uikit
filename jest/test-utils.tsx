import React from 'react';
import { render, RenderResult, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';

type ProvidersWrapperProps = {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    theme: any;
};
const ProvidersWrapper: React.FC<ProvidersWrapperProps> = ({
    theme,
    children,
}) => {
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

interface CustomRenderOptions extends RenderOptions {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    theme: any;
}

type CustomRender = (
    Component: React.ReactElement,
    options: CustomRenderOptions
) => RenderResult;

const customRender: CustomRender = (Component, { theme, ...options }) => {
    return render(
        <ProvidersWrapper theme={theme}>{Component}</ProvidersWrapper>,
        options
    );
};

export * from '@testing-library/react';
export { customRender as render };
