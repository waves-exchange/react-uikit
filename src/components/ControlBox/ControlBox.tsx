import styled, { CSSObject } from '@emotion/styled';
import css from '@styled-system/css';
import { Flex } from '../Flex/Flex';

export interface IControlBoxStyles {
    baseStyles?: CSSObject;
    _focus?: CSSObject;
    _hover?: CSSObject;
    _invalid?: CSSObject;
    _disabled?: CSSObject;
    _checked?: CSSObject;
    _child?: CSSObject;
    _checkedAndChild?: CSSObject;
    _checkedAndDisabled?: CSSObject;
    _checkedAndDisabledHover?: CSSObject;
    _checkedAndFocus?: CSSObject;
    _checkedAndHover?: CSSObject;
}

type TProps = IControlBoxStyles & {
    type?: 'checkbox' | 'radio';
};

export const ControlBox = styled(Flex)<TProps>(
    ({
        type = 'checkbox',
        _hover,
        _invalid,
        _disabled,
        _focus,
        _checked,
        _child = { opacity: 0 },
        _checkedAndChild = { opacity: 1 },
        _checkedAndDisabled,
        _checkedAndDisabledHover,
        _checkedAndFocus,
        _checkedAndHover,
    }: TProps) => {
        /* eslint-disable max-len */
        const checkedAndDisabled = `input[type=${type}]:checked:disabled + &, input[type=${type}][aria-checked=mixed]:disabled + &`;
        const checkedAndHover = `input[type=${type}]:checked:hover:not(:disabled) + &, input[type=${type}][aria-checked=mixed]:hover:not(:disabled) + &`;
        const checkedAndFocus = `input[type=${type}]:checked:focus + &, input[type=${type}][aria-checked=mixed]:focus + &`;
        const disabled = `input[type=${type}]:disabled + &`;
        const focus = `input[type=${type}]:focus + &`;
        const hover = `input[type=${type}]:hover:not(:disabled):not(:checked) + &`;
        const checkedAndDisabledHover = `input[type=${type}]:hover:disabled:checked + &`;
        const checked = `input[type=${type}]:checked + &, input[type=${type}][aria-checked=mixed] + &`;
        const invalid = `input[type=${type}][aria-invalid=true] + &`;

        return css({
            [focus]: _focus,
            [hover]: _hover,
            [disabled]: _disabled,
            [invalid]: _invalid,
            [checkedAndDisabled]: _checkedAndDisabled,
            [checkedAndFocus]: _checkedAndFocus,
            [checkedAndHover]: _checkedAndHover,
            [checkedAndDisabledHover]: _checkedAndDisabledHover,
            ...{ '& > *': _child },
            [checked]: {
                ..._checked,
                '& > *': _checkedAndChild,
            },
        });
    }
);

ControlBox.displayName = 'ControlBox';

ControlBox.defaultProps = {
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'center',
    flexShrink: 0,
    'aria-hidden': 'true',
};
