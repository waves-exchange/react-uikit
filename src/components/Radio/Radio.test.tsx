import React, { FC } from 'react';
import { render, fireEvent } from 'test-utils';
import {
    Radio,
    radioTestId,
    radioInputTestId,
    radioControlBoxTestId,
} from './Radio';
import { IControlBoxStyles } from '../ControlBox/ControlBox';

describe('<Radio>', () => {
    const label = 'label';

    it('becomes checked on click', () => {
        const { getByLabelText, getByTestId } = render(<Radio>{label}</Radio>);

        const inputElem = getByTestId(radioInputTestId);

        expect(inputElem).not.toBeChecked();

        fireEvent.click(getByLabelText(label));

        expect(inputElem).toBeChecked();
    });

    it('merges default and passed control box styles', () => {
        const controlBoxStyles: IControlBoxStyles = {
            baseStyles: { width: 40, height: 40 },
            _hover: { backgroundColor: 'red' },
            _checked: { '::before': { backgroundColor: 'red' } },
            _disabled: { backgroundColor: 'red' },
            _checkedAndDisabled: { '::before': { backgroundColor: 'red' } },
        };
        const { getByTestId } = render(
            <Radio controlBoxStyles={controlBoxStyles}>{label}</Radio>
        );

        expect(getByTestId(radioControlBoxTestId)).toMatchSnapshot();
    });

    it('renders custom control box', () => {
        const customControlBoxText = 'test';

        const CustomControl: FC = () => <div>{customControlBoxText}</div>;

        const { getByText } = render(
            <Radio customControlBox={true}>
                <CustomControl />
            </Radio>
        );

        getByText(customControlBoxText);
    });

    describe('prop: "onChange"', () => {
        it('calls if readOnly === true', () => {
            const onChange = jest.fn();

            const { getByLabelText } = render(
                <Radio onChange={onChange} readOnly={false}>
                    {label}
                </Radio>
            );

            fireEvent.click(getByLabelText(label));

            expect(onChange).toHaveBeenCalledTimes(1);
        });

        it("doesn't call if readOnly prop === true", () => {
            const onChange = jest.fn();

            const { getByLabelText } = render(
                <Radio onChange={onChange} readOnly={true}>
                    {label}
                </Radio>
            );

            fireEvent.click(getByLabelText(label));

            expect(onChange).not.toHaveBeenCalled();
        });
    });

    describe('becomes checked when passed props', () => {
        it('readOnly !== true && defaultChecked === true', () => {
            const { getByTestId } = render(
                <Radio defaultChecked={true}>{label}</Radio>
            );

            expect(getByTestId(radioInputTestId)).toBeChecked();
        });
        it('checked === true', () => {
            const { getByTestId } = render(
                <Radio checked={true} onChange={jest.fn()}>
                    {label}
                </Radio>
            );

            expect(getByTestId(radioInputTestId)).toBeChecked();
        });
    });

    describe('becomes unchecked when passed props', () => {
        it('checked === false', () => {
            const { getByTestId } = render(
                <Radio checked={false} onChange={jest.fn()}>
                    {label}
                </Radio>
            );

            expect(getByTestId(radioInputTestId)).not.toBeChecked();
        });
    });

    it('forwards style props', () => {
        const marginTop = 40;

        const { getByTestId } = render(<Radio mt={marginTop}>{label}</Radio>);

        expect(getByTestId(radioTestId)).toHaveStyle(
            `margin-top: ${marginTop}px`
        );
    });
});
