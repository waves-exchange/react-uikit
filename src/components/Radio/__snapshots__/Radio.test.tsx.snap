// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio> merges default and passed control box styles 1`] = `
.emotion-0 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 40px;
  height: 40px;
  margin-right: 4px;
  box-sizing: border-box;
  border: 1px solid;
  border-radius: circle;
  background-color: main.$300;
  border-color: main.$300;
  cursor: pointer;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0::before {
  content: "";
  display: block;
  -webkit-transition: 0.3s;
  transition: 0.3s;
  border-radius: circle;
  width: 12px;
  height: 12px;
  background-color: standard.$0;
  border: 2px solid;
  border-color: standard.$0;
}

input[type=radio]:hover:not(:disabled):not(:checked) + .emotion-0 {
  background-color: red;
  border-color: main.$200;
}

input[type=radio]:disabled + .emotion-0 {
  background-color: red;
  border-color: main.$600;
  cursor: default;
}

input[type=radio]:checked:disabled + .emotion-0::before,
input[type=radio][aria-checked=mixed]:disabled + .emotion-0::before {
  background-color: red;
}

.emotion-0 > * {
  opacity: 0;
}

input[type=radio]:checked + .emotion-0::before,
input[type=radio][aria-checked=mixed] + .emotion-0::before {
  background-color: red;
}

input[type=radio]:checked + .emotion-0 > *,
input[type=radio][aria-checked=mixed] + .emotion-0 > * {
  opacity: 1;
}

<div
  aria-hidden="true"
  class="emotion-0"
  data-testid="radio-control-box"
  type="radio"
/>
`;
